src/server/worker/pipeline/transform/pending_job.go

```diff
@@ -293,111 +261,119 @@ func (pj *pendingJob) createJobDatumFileSetParallel(ctx context.Context, taskDoe
 	// There are no serial datums if no base exists.
 	if pj.baseMetaCommit == nil {
-		return nil
+		return datum.CreateEmptyFileSet(pachClient)
 	}
-	pachClient := pj.driver.PachClient().WithCtx(ctx)
 	// Wait for the base job to finish.
 	ci, err := pachClient.WaitCommit(pj.baseMetaCommit.Branch.Repo.Name, pj.baseMetaCommit.Branch.Name, pj.baseMetaCommit.ID)
 	if err != nil {
-		return err
+		return "", err
 	}
 	if ci.Error != "" {
-		return pfsserver.ErrCommitError{Commit: ci.Commit}
+		return "", pfsserver.ErrCommitError{Commit: ci.Commit}
 	}
-	return pachClient.WithRenewer(func(ctx context.Context, renewer *renew.StringSet) error {
+	var outputFileSetID string
+	if err := pachClient.WithRenewer(func(ctx context.Context, renewer *renew.StringSet) error {
 		// Upload the datums from the current job into the datum file set format.
-		fileSetID, _, err := pj.createFullJobDatumFileSet(ctx, taskDoer, renewer)
-		if err != nil {
-			return err
-		}
-		// Create the output datum file set for the datums that were not processed by the base (failed, recovered, etc.).
-		outputFileSetID, deleteFileSetID, skipped, err := pj.createJobDatumFileSetSerial(ctx, taskDoer, renewer, fileSetID, pj.baseMetaCommit)
-		if err != nil {
-			return err
-		}
-		// Delete the appropriate files.
-		if err := pj.logger.LogStep("deleting old datum outputs", func() error {
-			pachClient := pachClient.WithCtx(ctx)
-			return pj.withDeleter(pachClient, func(deleter datum.Deleter) error {
-				dit := datum.NewFileSetIterator(pachClient, deleteFileSetID)
-				return errors.EnsureStack(dit.Iterate(func(meta *datum.Meta) error {
-					return deleter(meta)
-				}))
-			})
+		var fileSetID string
+		if err := pj.logger.LogStep("creating full job datum file set", func() error {
+			fileSetID, err = createDatums(pachClient, taskDoer, pj.ji.Job)
+			if err != nil {
+				return err
+			}
+			return renewer.Add(ctx, fileSetID)
 		}); err != nil {
 			return errors.EnsureStack(err)
 		}
-		// Record the skipped datums.
-		stats := &datum.Stats{ProcessStats: &pps.ProcessStats{}}
-		stats.Skipped = skipped
-		pj.saveJobStats(stats)
-		if err := pj.writeJobInfo(); err != nil {
-			return err
-		}
-		return cb(ctx, outputFileSetID)
-	})
+		// Create the output datum file set for the datums that were not processed by the base (failed, recovered, etc.).
+		outputFileSetID, err = pj.createJobDatumFileSetSerial(ctx, taskDoer, renewer, fileSetID, pj.baseMetaCommit)
+		return err
+	}); err != nil {
+		return "", err
+	}
+	return outputFileSetID, nil
 }
 
-func (pj *pendingJob) createJobDatumFileSetSerial(ctx context.Context, taskDoer task.Doer, renewer *renew.StringSet, fileSetID string, baseMetaCommit *pfs.Commit) (string, string, int64, error) {
-	var outputFileSetID, deleteFileSetID string
-	var skipped int64
+func (pj *pendingJob) createJobDatumFileSetSerial(ctx context.Context, taskDoer task.Doer, renewer *renew.StringSet, fileSetID string, baseMetaCommit *pfs.Commit) (string, error) {
```

Comment:
not something to address in this PR, but I think it might be worth computing the serial + parallel datums in the same pass of the datums

Reply:
I have considered making the parallel and serial steps one stream where the parallel are produced immediately and the serial are produced when the base commit is finished. We could also just parallelize both steps rather than running them serially. I am going to do neither though because we are going to remove parallel jobs, which means only the serial step, with some modifications, will run.